### PR TITLE
Improve FEEC support for multi-patch domains

### PR DESCRIPTION
**Pull request by saidctb**
the following changes have been made:
- Add a DiscreteDerhamMultipatch class consisting of broken (single-patch) deRham sequences and broken P and D operators
- Add a conforming Projection for the multipatch V0 space 
- Add an example of a Conga Poisson solver on a multipatch domain (with mapping).
- Fix bug for field evaluation in the case of compatible spaces.
- Allow nested blocks in the BlockVector and BlockMatrix classes
- Improve `array_to_stencil` function and test it
- Fix bugs in matrix allocation for system of equations (serial/parallel, single/multi-patch)
- Add 2D time-harmonic Maxwell example (curl-curl)

### CODE REVIEW

File: psydac/api/essential_bc.py [hunk 1]
```diff
@@ -7,23 +7,24 @@
 
 
 #==============================================================================
-def apply_essential_bc_1d_StencilVector(V, bc, a):
+def apply_essential_bc_1d_StencilVector(V, a, bc=None, axis=None, ext=None, order=0):
```
Comment (yguclu) [inline psydac/api/essential_bc.py:10]: What does a default value of `None` mean for the `bc` argument? I don't think the code can work with that value

Comment (yguclu) [reply psydac/api/essential_bc.py:10]: Now I see, you get the information from `bc` if provided, and otherwise you get it from `axis`, `ext` and `order`. Why do you want to give this flexibility to these low-level functions?

File: psydac/feec/multipatch/operators.py [hunk 2]
```diff
@@ -116,32 +130,72 @@ def __init__(self, V0h, domain_h, homogeneous_bc=False, kappa=1e10, tol=1e-12, v
         ah = discretize(a, domain_h, [V0h, V0h])    # ... or (V0h, V0h)?
 
         self._A = ah.assemble() #.toarray()
```
Comment (yguclu) [inline psydac/feec/multipatch/operators.py]: I think that lines 115 to 132 should only be processed if `explicit == False`, correct?

File: psydac/feec/multipatch/operators.py [hunk 3]
```diff
@@ -116,32 +130,72 @@ def __init__(self, V0h, domain_h, homogeneous_bc=False, kappa=1e10, tol=1e-12, v
...
+            self._A[0,0][:,e1,0,0] = 1/2
+            self._A[1,1][:,s2,0,0] = 1/2
+
+            self._A[0,1][:,d1,0,-d2] = 1/2
+            self._A[1,0][:,s2,0, d1] = 1/2
+
+            for b in domain.boundary:
+                i = get_space_indices_from_target(domain, b)
+                for j in range(len(domain)):
+                    apply_essential_bc(V0h.spaces[i], self._A[i,j], axis=b.axis, ext=b.ext)
```
Comment (yguclu) [inline psydac/feec/multipatch/operators.py:165]: I think that these boundary conditions should only be imposed if `homogeneous_bc == True`.

**Review by yguclu — CHANGES_REQUESTED** [review]
I am halfway through reviewing this PR. These are my comments so far.

File: psydac/api/discretization.py [hunk 4]
```diff
@@ -270,192 +268,91 @@ def solve(self, **kwargs):
...
+        if isinstance(V.kind, (H1SpaceType, L2SpaceType, UndefinedSpaceType)):
+            Vh = ProductFemSpace(*[Vh]*V.ldim)
 
-    return DiscreteDerham(mapping, *d_spaces)
+    return Vh
 
 #==============================================================================
-# TODO multi patch
 # TODO knots
 def discretize_space(V, domain_h, *args, **kwargs):
```
Comment (yguclu) [inline psydac/api/discretization.py:331]: Can you explain a bit how the algorithm that is used in this function? It is hard to follow without any comment in the code.

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/api/tests/test_api_feec_2d.py [hunk 5]
```diff
@@ -282,8 +282,8 @@ class CollelaMapping2D(Mapping):
     derham_h = discretize(derham, domain_h, degree=[degree, degree], periodic=[periodic, periodic])
 
     # Discrete bilinear forms
-    a1_h = discretize(a1, domain_h, (derham_h.V1, derham_h.V1), backend=PSYDAC_BACKEND_GPYCCEL)
-    a2_h = discretize(a2, domain_h, (derham_h.V2, derham_h.V2), backend=PSYDAC_BACKEND_GPYCCEL)
+    a1_h = discretize(a1, domain_h, (derham_h.V1, derham_h.V1)) #, backend=PSYDAC_BACKEND_GPYCCEL)
+    a2_h = discretize(a2, domain_h, (derham_h.V2, derham_h.V2)) #, backend=PSYDAC_BACKEND_GPYCCEL)
```
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/tests/test_api_feec_2d.py:286]: they were needed on my machine, but I don't think they should be kept in the PR

File: psydac/api/tests/test_api_feec_2d.py [hunk 6]
```diff
@@ -294,7 +294,7 @@ class CollelaMapping2D(Mapping):
 
     # Discretize and assemble penalization matrix
     if not periodic:
-        a1_bc_h = discretize(a1_bc, domain_h, (derham_h.V1, derham_h.V1), backend=PSYDAC_BACKEND_GPYCCEL)
+        a1_bc_h = discretize(a1_bc, domain_h, (derham_h.V1, derham_h.V1)) #, backend=PSYDAC_BACKEND_GPYCCEL)
```
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/tests/test_api_feec_2d.py:297]: no, not needed on this PR

File: psydac/api/feec.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    """
+    def __init__(self, mapping, *spaces):
+
+        dim           = len(spaces) - 1
+        self._dim     = dim
+        self._spaces  = spaces
+        self._mapping = mapping
+
+        if dim == 1:
+            D0 = Derivative_1D(spaces[0], spaces[1])
```
Comment (campospinto) [inline psydac/api/feec.py:25]: is there a verification in the construction of D0 that the spaces form a sequence ? (same question for the other diff operators)

File: psydac/fem/vector.py [hunk 8]
```diff
@@ -254,3 +264,6 @@ def n_components( self ):
     @property
     def comm( self ):
         return self.spaces[0].comm
+
+    def __len__(self):
+        return len(self.spaces)
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/fem/vector.py:269]: we have the same thing for the class domain in Sympde, and it's more readable for me, so I prefer to add a `__getitem__` then removing it

File: psydac/api/grid.py [hunk 9]
```diff
@@ -157,6 +157,9 @@ def __init__( self, V, nderiv , trial=False, grid=None, ext=None):
                 points = grid.points[axis]
                 boundary_basis = basis_ders_on_quad_grid(
                         space.knots, space.degree, points, nderiv, space.basis)
+                if space.basis == 'M':
+                    import numpy as np
+                    boundary_basis = np.zeros_like(boundary_basis)
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/grid.py:162]: some components of the space are zero when we are in the boundary, and this ensures it.
this is something we didn't agree on yet, so I added the hack outside of the  `basis_ders_on_quad_grid`

File: psydac/fem/basic.py [hunk 10]
```diff
@@ -186,7 +186,7 @@ def __init__( self, space, coeffs=None ):
 
         if coeffs is not None:
             assert isinstance( coeffs, Vector )
-            assert space.vector_space is coeffs.space
+            assert space.vector_space == coeffs.space
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/fem/basic.py:189]: I don't remember anymore, maybe it used maybe not

File: psydac/linalg/utilities.py [hunk 11]
```diff
@@ -12,14 +12,26 @@ def array_to_stencil(x, Xh):
...
+
+        else:
+            starts = [np.array(V.starts) for V in Xh.spaces]
+            ends   = [np.array(V.ends)   for V in Xh.spaces]
+
+            for i in range(len(starts)):
+                g = tuple(slice(s,e+1) for s,e in zip(starts[i], ends[i]))
+                shape = tuple(ends[i]-starts[i]+1)
+                u[i][g] = x[:np.product(shape)].reshape(shape)
+                x       = x[np.product(shape):]
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/utilities.py:34]: in the stockes equation for example, if you want to try different solvers

File: psydac/ddm/cart.py [hunk 12]
```diff
@@ -222,6 +222,11 @@ def shape( self ):
     def subcomm( self ):
         return self._subcomm
 
+    def __eq__( self, a):
+        a = (a.npts, a.pads, a.periods, a.comm)
+        b = (self.npts, self.pads, self.periods, self.comm)
+        return a == b
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/ddm/cart.py:228]: we need when we compare two spaces , we need to make sure that the carts are the same

File: psydac/api/ast/fem.py [hunk 13]
```diff
@@ -527,7 +527,8 @@ def _create_ast_bilinear_form(terminal_expr, atomic_expr_field,
         args['f_coeffs']       = flatten(list(g_coeffs.values()))
         args['field_basis']    = tuple(d_fields[f]['global'] for f in fields)
         args['fields_degrees'] = lengths_fields.values()
-        args['fields']         = fields
+        fields                 = tuple(f.base if isinstance(f, IndexedVectorFunction) else f for f in fields)
+        args['fields']         = tuple(dict.fromkeys(fields))
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/ast/fem.py:531]: yes

File: psydac/api/feec.py (hunk 7 above)
Comment (saidctb) [reply psydac/api/feec.py:25]: no, I don't think that we verify that

File: psydac/fem/vector.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/fem/vector.py:269]: Fine with me!

File: psydac/api/grid.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/grid.py:162]: I see, then please do these two things:

1. Add one (or more) line(s) of comment before this three lines of code, where you explain with a bit of detail what you are doing and why;
2. Instead of using `zeros_like` from `numpy`, you can just set the array to zero:
```python
if space.basis == 'M':
    boundary_basis[...] = 0.0

[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/grid.py:162]: By the way, `basis_ders_on_quad_grid` computes the values of the 1D basis functions, so this "hack" is definitely not related to it.

And I am still not clear why a component of the space should be zero on the boundary, unless we are using homogeneous Dirichlet boundary conditions. Can you give us an example?

File: psydac/api/ast/fem.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/ast/fem.py:531]: I see, thanks

File: psydac/linalg/utilities.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/linalg/utilities.py:34]: Good. Then this function needs some unit tests. Above I sketched how they should look like.

File: psydac/api/feec.py (hunk 7 above)
Comment (yguclu) [reply psydac/api/feec.py:25]: @campospinto It is not trivial to check that the spaces form a sequence, we could do that but it is quite a lot of code. 

Here we assume that they do, because these differential operators are only used by class `DiscreteDerham`, which is only created by function `discretize_derham`, which creates the appropriate spaces as part of a de Rham complex.

We could reorganize the whole process in the future. Suggestions are welcome!

File: psydac/api/grid.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/grid.py:162]: if you take a unit square domain, the `Hcurl` space on the boundary `x = 1` is equal to `(0, N(x=1)*M)`

File: psydac/fem/vector.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/fem/vector.py:269]: For readability I think that writing len(V.spaces) is better than len(V). The former is not much longer and clearly denotes the number of spaces in the product (and also makes clear that it is a product of spaces). The latter makes an implicit assumption of the type of V (which could be misleading eg in the case of a vector FEM space defined as a product of scalar FEM spaces...) and could also mean the length of the space basis, ie its dimension.

File: psydac/api/expr.py [hunk 14]
```diff
@@ -126,15 +125,14 @@ def __call__(self, *args, **kwargs):
                     spans   = list(map(list, zip(*spans)))
                     basis   = [b for bs in basis for b in bs]
                     spans   = [s for sp in spans for s in sp]
-                    coeffs  = coeffs + tuple(F.coeffs[i] for i in range(Vh.shape))
```
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/expr.py:129]: I agree, for the same reasons as above.

File: psydac/api/grid.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/grid.py:162]: But as Yaman was pointing, this is only true for a function in H_0(curl), not for a general H(curl) function.

File: psydac/api/ast/expr.py [hunk 15]
```diff
@@ -196,7 +196,7 @@ def _initialize(self):
         expr = self.expr
         dim  = Vh.ldim
         if isinstance(Vh, ProductFemSpace):
-            size = Vh.shape
```
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/ast/expr.py:199]: I agree (see my comment on fem/vector.py)

File: psydac/fem/vector.py [hunk 16]
```diff
@@ -156,8 +155,19 @@ class ProductFemSpace( FemSpace ):
...
+            return args[0]
+        else:
+            return FemSpace.__new__(cls)
+
     def __init__( self, *args, **kwargs ):
         """."""
+
+        if len(args) == 1:
+            return
+
```
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/fem/vector.py:170]: Here I think we are in the class `ProductFemSpace`. But does that mean that in the case of a single argument, the constructor returns an object which will not be a ProductFemSpace ?

File: psydac/api/grid.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/grid.py:162]: By the way, the docstring of the `BasisValues` class should be more detailed. 

In particular, it should specify on which grid the values are computed,
and also probably indicate that the derivatives of the basis values are computed there (as it seems). 

If the parameter 'grid' is only for the boundaries, it should probably be called boundary_grid.

File: psydac/api/feec.py (hunk 7 above)
Comment (campospinto) [reply psydac/api/feec.py:25]: Ok ! Thanks for clarifying

File: psydac/fem/vector.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/fem/vector.py:170]: I assumed so, but we do for some reason

File: psydac/api/essential_bc.py [hunk 17]
```diff
@@ -120,6 +120,7 @@ def apply_essential_bc_BlockVector(a, bc):
     if bc.index_component:
         for i_loc in bc.index_component:
             i = bc.position + i_loc
+
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/essential_bc.py]: done

File: psydac/api/ast/expr.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/ast/expr.py:199]: done

File: psydac/fem/vector.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/fem/vector.py]: done

File: psydac/api/tests/test_api_feec_1d.py [hunk 18]
```diff
@@ -162,6 +162,7 @@ class CollelaMapping1D(Mapping):
             from psydac.api.essential_bc import apply_essential_bc
             M0_dir   = M0.copy()
             D0_T_dir = D0_T.copy()
+
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/tests/test_api_feec_1d.py]: done

File: psydac/api/expr.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/expr.py:129]: done

File: psydac/api/grid.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/grid.py:162]: yes,it's when we have homogenous boundary conditions

File: psydac/fem/vector.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/fem/vector.py:170]: We do what? For which reason ?

File: psydac/api/grid.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/grid.py:162]: so in this case the values on the boundary should be zero. If there are not, it's because of an error somewhere else. This hack may just hide that.

[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/grid.py]: yes we set them here, there is no other place

File: psydac/linalg/utilities.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/utilities.py]: done

File: psydac/api/grid.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/grid.py]: I have deleted it, untill we properly implement homogenous boundary conditions

File: psydac/linalg/tests/test_stencil_vector.py [hunk 19]
```diff
@@ -138,6 +139,25 @@ def test_stencil_vector_2d_serial_dot( n1, n2, p1, p2, P1=True, P2=False ):
...
+    x = StencilVector( V )
+
+    for i1 in range(n1):
+        for i2 in range(n2):
+            x[i1,i2] = 10*i1 + i2
+
+    xa = x.toarray()
+    v  = array_to_stencil(xa, V)
+
+    assert np.allclose( xa , v.toarray() )
```
Comment (yguclu) [inline psydac/linalg/tests/test_stencil_vector.py:159]: @saidctb Thanks for adding this test. Can you please parametrize `P1` and `P2` as well?

File: psydac/linalg/tests/test_block.py [hunk 20]
```diff
@@ -170,6 +171,37 @@ def test_block_matrix( n1, n2, p1, p2, P1, P2  ):
...
+    for i1 in range(n1):
+        for i2 in range(n2):
+            x[0][i1,i2] = 2.0*random() + 1.0
+            x[1][i1,i2] = 5.0*random() - 1.0
+    x.update_ghost_regions()
+
+    xa = x.toarray()
+    v  = array_to_stencil(xa, W)
+
+    assert np.allclose( xa , v.toarray() )
```
Comment (yguclu) [inline psydac/linalg/tests/test_block.py:202]: I think that we also need to test the case where one of the two spaces `V1, V2` is a `BlockVectorSpace` itself. This is relevant to our multi-patch vector examples if I am not mistaken.

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/fem/vector.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/fem/vector.py:170]: @campospinto You're right of course! This is class `ProductFemSpace`. In the linear algebra layer class `BlockVectorSpace` returns the input `VectorSpace` if only one is provided (see [here](https://github.com/pyccel/psydac/blob/79392da8a71f0c1aeaa11ccc48158c7aaaf3f37c/psydac/linalg/block.py#L14-L39)), so it is natural to do the same in the FEM layer.

[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/fem/vector.py:170]: @saidctb Can you please read my original comment (the first at the top of this discussion) and reply to it?

File: psydac/linalg/utilities.py [hunk 21]
```diff
@@ -32,4 +44,46 @@ def array_to_stencil(x, Xh):
...
+        ends   = np.array(Xh.ends)
+
+        g = tuple(slice(s,e+1) for s,e in zip(starts, ends))
+        shape = tuple(ends-starts+1)
+        u[g] = x[:np.product(shape)].reshape(shape)
+    else:
+        raise ValueError('Xh must be a StencilVectorSpace or a BlockVectorSpace')
+
+    u.update_ghost_regions()
     return u
```
Comment (yguclu) [inline psydac/linalg/utilities.py:89]: I think that you can write this function as:
```python
def petsc_to_stencil(x, Xh):
    assert isinstance(x, ...)
    return array_to_stencil(x.array, Xh)
```
So why duplicating so much code?

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/linalg/tests/test_stencil_vector.py (hunk 19 above)
Comment (saidctb) [reply psydac/linalg/tests/test_stencil_vector.py]: done

File: psydac/linalg/tests/test_block.py (hunk 20 above)
Comment (saidctb) [reply psydac/linalg/tests/test_block.py]: done

File: psydac/linalg/utilities.py (hunk 21 above)
Comment (saidctb) [reply psydac/linalg/utilities.py]: ah yes, you are right

File: psydac/fem/vector.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/fem/vector.py:170]: if you delete that line in __init__, the tests will not work anymore.
if you run the test `test_api_1d_compatible_spaces.py` you will get this
```python
test_api_1d_compatible_spaces.py:55: in run_system_1_1d_dir
    ah = discretize(equation, domain_h, [Xh, Xh])
../discretization.py:498: in discretize
    return DiscreteEquation(a, *args, **kwargs)
../discretization.py:114: in __init__
    self._lhs = discretize(expr.lhs, *newargs, **kwargs)
../discretization.py:489: in discretize
    return DiscreteBilinearForm(a, kernel_expr, *args, **kwargs)
../fem.py:157: in __init__
    quad_order                    = kwargs.pop('quad_order', get_quad_order(self.spaces[1]))
../fem.py:42: in get_quad_order
    return get_quad_order(Vh.spaces[0])
../fem.py:42: in get_quad_order
    return get_quad_order(Vh.spaces[0])
E   RecursionError: maximum recursion depth exceeded
!!! Recursion detected (same locals & position)
```
the problem with that is even if `len(args) == 1`  it will still pass by the `__init___`

File: psydac/fem/vector.py [hunk 22]
```diff
@@ -165,10 +165,10 @@ def __new__(cls, *spaces):
     def __init__( self, *spaces):
         """."""
 
-        if len(args) == 1:
+        if len(spaces) == 1:
             return
 
-        self._spaces = tuple(args)
+        self._spaces = tuple(spaces)
```
Comment (yguclu) [inline psydac/fem/vector.py:171]: By construction `spaces` is already a `tuple`, which is an immutable object. So there is no need to create a new tuple here, and the following suffices:
```python
self._spaces = spaces
```

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/linalg/tests/test_block.py (hunk 20 above)
Comment (yguclu) [reply psydac/linalg/tests/test_block.py]: Thanks

File: psydac/linalg/tests/test_stencil_vector.py (hunk 19 above)
Comment (yguclu) [reply psydac/linalg/tests/test_stencil_vector.py]: Thanks

File: psydac/fem/vector.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/fem/vector.py]: This is really surprising! According to the [Python documentation](https://docs.python.org/3/reference/datamodel.html#object.__new__):

> If __new__() does not return an instance of cls, then the new instance’s __init__() method will not be invoked.

If in this case the `__init__()` method is invoked, it means that `args[0]` is an object of type `ProductFemSpace`. I don't see why this should happen, and I wonder if we have a bug somewhere else.

[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/fem/vector.py]: yes I have zero ideas, why it's doing that

File: psydac/fem/vector.py (hunk 22 above)
Comment (saidctb) [reply psydac/fem/vector.py]: done

File: psydac/fem/vector.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/fem/vector.py]: I've just checked, the problem is [here](https://github.com/pyccel/psydac/blob/d9b19a86f86f2c60ed38ef3cdedb8e49e38807b7/psydac/api/discretization.py#L444):
```python
Vh = ProductFemSpace(*g_spaces.values())
```
when `g_spaces.values()` could contain just one object of type `ProductFemSpace`.

[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/fem/vector.py]: @saidctb I think that this is not a bug, just a strange way to organize the algorithm. Your code seems OK to me, thanks!

File: psydac/api/discretization.py (hunk 4 above)
Comment (saidctb) [reply psydac/api/discretization.py:331]: okay, the discretize_space is divided into two cases, the case where we have a geometry file and the case where we have either an analytical mapping or our domain is the logical domain.
In the two cases, we build the dictionary `g_spaces` that contains the interiors as keys and the spaces as value, In the case of single patch , we have one key and one value

Comment (saidctb) [reply psydac/api/discretization.py:331]: after building the `g_spaces` we create the compatible spaces if needed in each interior domain

Comment (yguclu) [reply psydac/api/discretization.py:331]: Thanks! Can you add a nice docstring to the function, with this explanation for future reference?

Comment (saidctb) [reply psydac/api/discretization.py:331]: okay

File: psydac/fem/basic.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/fem/basic.py]: I've removed this with commit 0f80a46. All unit tests seem to work.

File: psydac/ddm/cart.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/ddm/cart.py]: I've commented out those lines of code with commit 0f80a46. All unit tests seem to work.

**Comment by yguclu** [comment]
@saidctb @campospinto This PR is almost ready. We need to add the iist of changes to this PR's description (just under the title). This will then be used as the message for the squashed merge commit.

File: psydac/api/discretization.py (hunk 4 above)
Comment (saidctb) [reply psydac/api/discretization.py:331]: done

**Review by campospinto — APPROVED** [review]

**Review by yguclu — APPROVED** [review]
